feat(docs): dynamic model and tool selection in create react agent (#5777)
langchain-ai/langgraph: docs/docs/agents/models.md

:::

:::python

### Dynamic model selection

Pass a callable function to `create_react_agent` to dynamically select the model at runtime. This is useful for scenarios where you want to choose a model based on user input, configuration settings, or other runtime conditions.

> The selector function must return an instance of a `BaseChatModel`. If you're using tools, you must bind the tools to the model within the selector function.

Author: ```suggestion
The selector function must return a chat model. If you're using tools, you must bind the tools to the model within the selector function.
```

  ```python
from dataclasses import dataclass
from typing import Literal
from langchain.chat_models import init_chat_model
from langchain_core.language_models import BaseChatModel
from langchain_core.tools import tool
from langgraph.prebuilt import create_react_agent
from langgraph.prebuilt.chat_agent_executor import AgentState
from langgraph.runtime import Runtime

@tool
def weather() -> str:
    """Returns the current weather conditions."""
    return "It's nice and sunny."

# Define the runtime context
@dataclass
class CustomContext:
    provider: Literal["anthropic", "openai"]

# Initialize models
openai_model = init_chat_model("openai:gpt-4o")
anthropic_model = init_chat_model("anthropic:claude-sonnet-4-20250514")

# Selector function for model choice

> def select_model(state: AgentState, runtime: Runtime[CustomContext]) -> BaseChatModel:

Author: ```suggestion
def select_model(state: AgentState, runtime: Runtime[CustomContext]):
```
